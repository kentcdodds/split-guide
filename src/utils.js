/* eslint no-console:0 */

export {getErrorLogger, getThenLogger}

function getErrorLogger(title) {
  return err => {
    console.error(`There was an error: ${title}`, err)
    return Promise.reject(err)
  }
}

function getThenLogger(title) {
  return res => {
    console.error(title, res)
    return res
  }
}
