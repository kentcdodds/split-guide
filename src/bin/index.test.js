import path from 'path'
import fs from 'fs'
import spawn from 'spawn-command'
import pify from 'pify'
import glob from 'glob'
import dirTree from 'directory-tree'
import yargsParser from 'yargs-parser'
import {getErrorLogger} from '../utils'

// this is a bit of a longrunning test...
jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000 // eslint-disable-line no-undef

const SPLIT_GUIDE_PATH = require.resolve('./index')
const BABEL_BIN_PATH = require.resolve('babel-cli/bin/babel-node')

test('split-guide --help', () => {
  return runSplitGuideCLI('--help').then(stdout => {
    expect(stdout).toMatchSnapshot()
  })
})

test('split-guide generate', () => {
  return runCLIAndAssertFileOutput(
    'generate',
    path.resolve(__dirname, '../../test/fixtures/generate'),
  )
})

test('split-guide generate --templates-dir guides --exercises-dir app --exercises-final-dir app-finished', () => {
  return runCLIAndAssertFileOutput(
    'generate --templates-dir guides --exercises-dir app --exercises-final-dir app-finished',
    path.resolve(__dirname, '../../test/fixtures/generate-with-args'),
  )
})

test('split-guide generate --no-clean=true --exercises-dir src --exercises-final-dir src-final', () => {
  return runCLIAndAssertFileOutput(
    'generate --no-clean=true --exercises-dir src --exercises-final-dir src-final',
    path.resolve(__dirname, '../../test/fixtures/generate-no-clean'),
  )
})

test('split-guide generate --ignore "**/*.ignore-me.js"', () => {
  return runCLIAndAssertFileOutput(
    'generate --ignore "**/*.ignore-me.js"',
    path.resolve(__dirname, '../../test/fixtures/generate-ignore-one'),
  )
})

test('split-guide generate --ignore "**/*.ignore-me.js" "**/*.no-copy.js"', () => {
  return runCLIAndAssertFileOutput(
    'generate --ignore "**/*.ignore-me.js" "**/*.no-copy.js"',
    path.resolve(__dirname, '../../test/fixtures/generate-ignore-multiple'),
  )
})

function runCLIAndAssertFileOutput(args, cwd) {
  const {exercisesDir = './exercises', exercisesFinalDir = './exercises-final'} = yargsParser(args)
  return runSplitGuideCLI(args, cwd).then(stdout => {
    expect(stdout).toMatchSnapshot()
    const tree = dirTree(cwd)
    relativeizePathInTree(tree)
    expect(tree).toMatchSnapshot()
    // cannot use Promise.all here because we need to make sure the snapshots are
    // taken in the correct order
    return expectDirectoryToMatchSnapshot(path.resolve(cwd, exercisesDir))
      .then(() => expectDirectoryToMatchSnapshot(path.resolve(cwd, exercisesFinalDir)))
  }, getErrorLogger('runSplitGuideCLI'))
}

function expectDirectoryToMatchSnapshot(directory) {
  return pify(glob)(path.resolve(directory, '**/*'), {nodir: true})
    .then(readAllFilesAsPromise)
    .then(expectFilesToMatchSnapshot)
    .catch(getErrorLogger(`expectDirectoryToMatchSnapshot(${directory})`))

  function readAllFilesAsPromise(files) {
    const allPromises = files.map(readFileAsPromise)
    return Promise.all(allPromises)
  }

  function readFileAsPromise(file) {
    return pify(fs.readFile)(file, 'utf8')
      .then(contents => ({file: relativeizePath(file), contents}))
      .catch(getErrorLogger(`readFileAsPromise(${file})`))
  }

  function expectFilesToMatchSnapshot(files) {
    expect(files).toMatchSnapshot()
  }
}

function runSplitGuideCLI(args = '', cwd = process.cwd()) {
  const isRelative = cwd[0] !== '/'
  if (isRelative) {
    cwd = path.resolve(__dirname, cwd)
  }

  return new Promise((resolve, reject) => {
    let stdout = ''
    let stderr = ''
    const command = `${BABEL_BIN_PATH} ${SPLIT_GUIDE_PATH} ${args}`
    const child = spawn(command, {cwd})

    child.on('error', error => {
      reject(error)
    })

    child.stdout.on('data', data => {
      stdout += data.toString()
    })

    child.stderr.on('data', data => {
      stderr += data.toString()
    })

    child.on('close', () => {
      if (stderr) {
        reject(stderr)
      } else {
        resolve(stdout)
      }
    })
  })
}

function relativeizePath(absolutePath) {
  return absolutePath.replace(path.resolve(__dirname, '../../'), '<projectRootDir>')
}

function relativeizePathInTree(tree) {
  tree.path = relativeizePath(tree.path)
  if (tree.children) {
    tree.children.forEach(relativeizePathInTree)
  }
}
