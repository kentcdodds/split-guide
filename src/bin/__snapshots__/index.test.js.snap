exports[`test split-guide --help 1`] = `
"Usage: src/bin/index.js

Commands:
  generate  generate your split-guide

Options:
  -h, --help  Show help  [boolean]

"
`;

exports[`test split-guide generate --ignore "**/*.ignore-me.js" "**/*.no-copy.js" 1`] = `""`;

exports[`test split-guide generate --ignore "**/*.ignore-me.js" "**/*.no-copy.js" 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple/exercises/README.md",
          "size": 97
        }
      ],
      "name": "exercises",
      "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple/exercises",
      "size": 97
    },
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple/exercises-final/README.md",
          "size": 97
        }
      ],
      "name": "exercises-final",
      "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple/exercises-final",
      "size": 97
    },
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple/templates/README.md",
          "size": 97
        },
        Object {
          "extension": ".js",
          "name": "foo.ignore-me.js",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple/templates/foo.ignore-me.js",
          "size": 46
        },
        Object {
          "extension": ".js",
          "name": "foo.no-copy.js",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple/templates/foo.no-copy.js",
          "size": 31
        }
      ],
      "name": "templates",
      "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple/templates",
      "size": 174
    }
  ],
  "name": "generate-ignore-multiple",
  "path": "<projectRootDir>/test/fixtures/generate-ignore-multiple",
  "size": 368
}
`;

exports[`test split-guide generate --ignore "**/*.ignore-me.js" "**/*.no-copy.js" 3`] = `
Array [
  Object {
    "contents": "# --ignore

This allows you to ignore files that match glob. And you can pass multiple of these!
",
    "file": "<projectRootDir>/test/fixtures/generate-ignore-multiple/exercises/README.md"
  }
]
`;

exports[`test split-guide generate --ignore "**/*.ignore-me.js" "**/*.no-copy.js" 4`] = `
Array [
  Object {
    "contents": "# --ignore

This allows you to ignore files that match glob. And you can pass multiple of these!
",
    "file": "<projectRootDir>/test/fixtures/generate-ignore-multiple/exercises-final/README.md"
  }
]
`;

exports[`test split-guide generate --ignore "**/*.ignore-me.js" 1`] = `""`;

exports[`test split-guide generate --ignore "**/*.ignore-me.js" 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-one/exercises/README.md",
          "size": 60
        }
      ],
      "name": "exercises",
      "path": "<projectRootDir>/test/fixtures/generate-ignore-one/exercises",
      "size": 60
    },
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-one/exercises-final/README.md",
          "size": 60
        }
      ],
      "name": "exercises-final",
      "path": "<projectRootDir>/test/fixtures/generate-ignore-one/exercises-final",
      "size": 60
    },
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-one/templates/README.md",
          "size": 60
        },
        Object {
          "extension": ".js",
          "name": "foo.ignore-me.js",
          "path": "<projectRootDir>/test/fixtures/generate-ignore-one/templates/foo.ignore-me.js",
          "size": 46
        }
      ],
      "name": "templates",
      "path": "<projectRootDir>/test/fixtures/generate-ignore-one/templates",
      "size": 106
    }
  ],
  "name": "generate-ignore-one",
  "path": "<projectRootDir>/test/fixtures/generate-ignore-one",
  "size": 226
}
`;

exports[`test split-guide generate --ignore "**/*.ignore-me.js" 3`] = `
Array [
  Object {
    "contents": "# --ignore

This allows you to ignore files that match glob
",
    "file": "<projectRootDir>/test/fixtures/generate-ignore-one/exercises/README.md"
  }
]
`;

exports[`test split-guide generate --ignore "**/*.ignore-me.js" 4`] = `
Array [
  Object {
    "contents": "# --ignore

This allows you to ignore files that match glob
",
    "file": "<projectRootDir>/test/fixtures/generate-ignore-one/exercises-final/README.md"
  }
]
`;

exports[`test split-guide generate --no-clean=true --exercises-dir src --exercises-final-dir src-final 1`] = `""`;

exports[`test split-guide generate --no-clean=true --exercises-dir src --exercises-final-dir src-final 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-no-clean/src/README.md",
          "size": 77
        },
        Object {
          "extension": ".js",
          "name": "persist.js",
          "path": "<projectRootDir>/test/fixtures/generate-no-clean/src/persist.js",
          "size": 56
        }
      ],
      "name": "src",
      "path": "<projectRootDir>/test/fixtures/generate-no-clean/src",
      "size": 133
    },
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-no-clean/src-final/README.md",
          "size": 54
        },
        Object {
          "extension": ".js",
          "name": "persist.js",
          "path": "<projectRootDir>/test/fixtures/generate-no-clean/src-final/persist.js",
          "size": 63
        }
      ],
      "name": "src-final",
      "path": "<projectRootDir>/test/fixtures/generate-no-clean/src-final",
      "size": 117
    },
    Object {
      "children": Array [
        Object {
          "extension": ".md",
          "name": "README.md",
          "path": "<projectRootDir>/test/fixtures/generate-no-clean/templates/README.md",
          "size": 178
        }
      ],
      "name": "templates",
      "path": "<projectRootDir>/test/fixtures/generate-no-clean/templates",
      "size": 178
    }
  ],
  "name": "generate-no-clean",
  "path": "<projectRootDir>/test/fixtures/generate-no-clean",
  "size": 428
}
`;

exports[`test split-guide generate --no-clean=true --exercises-dir src --exercises-final-dir src-final 3`] = `
Array [
  Object {
    "contents": "// this will not get deleted!
// isn't that cool!? ðŸ˜Ž
",
    "file": "<projectRootDir>/test/fixtures/generate-no-clean/src/persist.js"
  },
  Object {
    "contents": "## --no-clean

This allows you to generate your stuff into an existing file.
",
    "file": "<projectRootDir>/test/fixtures/generate-no-clean/src/README.md"
  }
]
`;

exports[`test split-guide generate --no-clean=true --exercises-dir src --exercises-final-dir src-final 4`] = `
Array [
  Object {
    "contents": "// this will not get deleted either!
// isn't that cool!? ðŸ˜Ž
",
    "file": "<projectRootDir>/test/fixtures/generate-no-clean/src-final/persist.js"
  },
  Object {
    "contents": "## --no-clean

Just don't forget to delete old files!
",
    "file": "<projectRootDir>/test/fixtures/generate-no-clean/src-final/README.md"
  }
]
`;

exports[`test split-guide generate --templates-dir guides --exercises-dir app --exercises-final-dir app-finished 1`] = `""`;

exports[`test split-guide generate --templates-dir guides --exercises-dir app --exercises-final-dir app-finished 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "<projectRootDir>/test/fixtures/generate-with-args/app/foo/bar/README.md",
                  "size": 65
                }
              ],
              "name": "bar",
              "path": "<projectRootDir>/test/fixtures/generate-with-args/app/foo/bar",
              "size": 65
            }
          ],
          "name": "foo",
          "path": "<projectRootDir>/test/fixtures/generate-with-args/app/foo",
          "size": 65
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "<projectRootDir>/test/fixtures/generate-with-args/app/root.js",
          "size": 92
        }
      ],
      "name": "app",
      "path": "<projectRootDir>/test/fixtures/generate-with-args/app",
      "size": 157
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "<projectRootDir>/test/fixtures/generate-with-args/app-finished/foo/bar/README.md",
                  "size": 61
                }
              ],
              "name": "bar",
              "path": "<projectRootDir>/test/fixtures/generate-with-args/app-finished/foo/bar",
              "size": 61
            },
            Object {
              "extension": ".md",
              "name": "final-only.md",
              "path": "<projectRootDir>/test/fixtures/generate-with-args/app-finished/foo/final-only.md",
              "size": 109
            }
          ],
          "name": "foo",
          "path": "<projectRootDir>/test/fixtures/generate-with-args/app-finished/foo",
          "size": 170
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "<projectRootDir>/test/fixtures/generate-with-args/app-finished/root.js",
          "size": 76
        }
      ],
      "name": "app-finished",
      "path": "<projectRootDir>/test/fixtures/generate-with-args/app-finished",
      "size": 246
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "<projectRootDir>/test/fixtures/generate-with-args/guides/foo/bar/README.md",
                  "size": 292
                }
              ],
              "name": "bar",
              "path": "<projectRootDir>/test/fixtures/generate-with-args/guides/foo/bar",
              "size": 292
            },
            Object {
              "extension": ".md",
              "name": "final-only.md",
              "path": "<projectRootDir>/test/fixtures/generate-with-args/guides/foo/final-only.md",
              "size": 232
            }
          ],
          "name": "foo",
          "path": "<projectRootDir>/test/fixtures/generate-with-args/guides/foo",
          "size": 524
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "<projectRootDir>/test/fixtures/generate-with-args/guides/root.js",
          "size": 261
        }
      ],
      "name": "guides",
      "path": "<projectRootDir>/test/fixtures/generate-with-args/guides",
      "size": 785
    }
  ],
  "name": "generate-with-args",
  "path": "<projectRootDir>/test/fixtures/generate-with-args",
  "size": 1188
}
`;

exports[`test split-guide generate --templates-dir guides --exercises-dir app --exercises-final-dir app-finished 3`] = `
Array [
  Object {
    "contents": "This is a fixture

This is workshop stuff

This is common stuff

",
    "file": "<projectRootDir>/test/fixtures/generate-with-args/app/foo/bar/README.md"
  },
  Object {
    "contents": "/* eslint-disable */
module.exports = function sum(a, b) {
  // return the sum of a and b
}
",
    "file": "<projectRootDir>/test/fixtures/generate-with-args/app/root.js"
  }
]
`;

exports[`test split-guide generate --templates-dir guides --exercises-dir app --exercises-final-dir app-finished 4`] = `
Array [
  Object {
    "contents": "This is a fixture

This is common stuff

This is final stuff
",
    "file": "<projectRootDir>/test/fixtures/generate-with-args/app-finished/foo/bar/README.md"
  },
  Object {
    "contents": "This will only go into the final directory
because the result is an empty string for the exercises directory
",
    "file": "<projectRootDir>/test/fixtures/generate-with-args/app-finished/foo/final-only.md"
  },
  Object {
    "contents": "/* eslint-disable */
module.exports = function sum(a, b) {
  return a + b
}
",
    "file": "<projectRootDir>/test/fixtures/generate-with-args/app-finished/root.js"
  }
]
`;

exports[`test split-guide generate 1`] = `""`;

exports[`test split-guide generate 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "<projectRootDir>/test/fixtures/generate/exercises/foo/bar/README.md",
                  "size": 65
                }
              ],
              "name": "bar",
              "path": "<projectRootDir>/test/fixtures/generate/exercises/foo/bar",
              "size": 65
            }
          ],
          "name": "foo",
          "path": "<projectRootDir>/test/fixtures/generate/exercises/foo",
          "size": 65
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "<projectRootDir>/test/fixtures/generate/exercises/root.js",
          "size": 92
        }
      ],
      "name": "exercises",
      "path": "<projectRootDir>/test/fixtures/generate/exercises",
      "size": 157
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "<projectRootDir>/test/fixtures/generate/exercises-final/foo/bar/README.md",
                  "size": 61
                }
              ],
              "name": "bar",
              "path": "<projectRootDir>/test/fixtures/generate/exercises-final/foo/bar",
              "size": 61
            },
            Object {
              "extension": ".md",
              "name": "final-only.md",
              "path": "<projectRootDir>/test/fixtures/generate/exercises-final/foo/final-only.md",
              "size": 109
            }
          ],
          "name": "foo",
          "path": "<projectRootDir>/test/fixtures/generate/exercises-final/foo",
          "size": 170
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "<projectRootDir>/test/fixtures/generate/exercises-final/root.js",
          "size": 76
        }
      ],
      "name": "exercises-final",
      "path": "<projectRootDir>/test/fixtures/generate/exercises-final",
      "size": 246
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "<projectRootDir>/test/fixtures/generate/templates/foo/bar/README.md",
                  "size": 292
                }
              ],
              "name": "bar",
              "path": "<projectRootDir>/test/fixtures/generate/templates/foo/bar",
              "size": 292
            },
            Object {
              "extension": ".md",
              "name": "final-only.md",
              "path": "<projectRootDir>/test/fixtures/generate/templates/foo/final-only.md",
              "size": 232
            }
          ],
          "name": "foo",
          "path": "<projectRootDir>/test/fixtures/generate/templates/foo",
          "size": 524
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "<projectRootDir>/test/fixtures/generate/templates/root.js",
          "size": 261
        }
      ],
      "name": "templates",
      "path": "<projectRootDir>/test/fixtures/generate/templates",
      "size": 785
    }
  ],
  "name": "generate",
  "path": "<projectRootDir>/test/fixtures/generate",
  "size": 1188
}
`;

exports[`test split-guide generate 3`] = `
Array [
  Object {
    "contents": "This is a fixture

This is workshop stuff

This is common stuff

",
    "file": "<projectRootDir>/test/fixtures/generate/exercises/foo/bar/README.md"
  },
  Object {
    "contents": "/* eslint-disable */
module.exports = function sum(a, b) {
  // return the sum of a and b
}
",
    "file": "<projectRootDir>/test/fixtures/generate/exercises/root.js"
  }
]
`;

exports[`test split-guide generate 4`] = `
Array [
  Object {
    "contents": "This is a fixture

This is common stuff

This is final stuff
",
    "file": "<projectRootDir>/test/fixtures/generate/exercises-final/foo/bar/README.md"
  },
  Object {
    "contents": "This will only go into the final directory
because the result is an empty string for the exercises directory
",
    "file": "<projectRootDir>/test/fixtures/generate/exercises-final/foo/final-only.md"
  },
  Object {
    "contents": "/* eslint-disable */
module.exports = function sum(a, b) {
  return a + b
}
",
    "file": "<projectRootDir>/test/fixtures/generate/exercises-final/root.js"
  }
]
`;
